CB-2534 Support sql server and oracle as internal db

oksana-cherniavskaia opened the pull request:
CB-2534 Support sql server and oracle as internal db

Commit:
CB-2534 Support sql server and oracle as internal db

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java:
```diff
@@ -42,10 +43,10 @@
 
     private static final String[][] TSQL_BEGIN_END_BLOCK = new String[][]{
-        {SQLConstants.BLOCK_BEGIN, SQLConstants.BLOCK_END}
+            {SQLConstants.BLOCK_BEGIN, SQLConstants.BLOCK_END}
```
Please fix your formatting configuration. The `continuation indent` should be `4`, you have `8`.

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java:
```diff
@@ -0,0 +1,71 @@
+     * @return the string
+     * @throws DBException the db exception
+     */
+    public static String translateScript(
```
Please specify nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java:
```diff
@@ -0,0 +1,71 @@
+            @NotNull SQLDialect sourceDialect,
+            @NotNull SQLDialect targetDialect,
+            @NotNull DBPPreferenceStore preferenceStore,
+            @NotNull String script) throws DBException {
```
```suggestion
            @NotNull String script
        ) throws DBException {
```

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+        this.sqlTranslateContext = sqlTranslateContext;
+    }
+
+    public List<? extends SQLScriptElement> translate(@NotNull SQLScriptElement element) throws DBException {
```
Please specify return type nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+
+    private SQLTranslateContext sqlTranslateContext;
+
+    public SQLQueryTranslator(SQLTranslateContext sqlTranslateContext) {
```
Please specify nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+        return Collections.singletonList(element);
+    }
+
+    private List<? extends SQLScriptElement> translateQuery(@NotNull SQLQuery query) {
```
Please specify return type nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox replied:
There's no return type annotation

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+     * @param statement the statement
+     * @return the list
+     */
+    protected List<? extends SQLScriptElement> translateStatement(@NotNull SQLQuery query,
+                                                                  @NotNull Statement statement) {
```
Please specify nullability using an appropriate annotation

```suggestion
    protected List<? extends SQLScriptElement> translateStatement(
        @NotNull SQLQuery query,
        @NotNull Statement statement
    ) {
```

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+                            newDataType = extendedDialect.getLargeNumericType();
+                        }
+                        break;
+                }
```
Please add an empty `default` branch to make it clear that no action should be performed.

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+                                    cd.getColumnSpecs().set(indexOf, extendedDialect.getAutoIncrementKeyword());
+                                }
+                                break;
+                        }
```
Same as previous comment

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+     *
+     * @return the sql translate context
+     */
+    public SQLTranslateContext getSqlTranslateContext() {
```
Please specify nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -14,47 +14,182 @@
+     *
+     * @param sqlTranslateContext the sql translate context
+     */
+    public void setSqlTranslateContext(SQLTranslateContext sqlTranslateContext) {
```
Please specify nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLTranslator.java:
```diff
@@ -0,0 +1,38 @@
+     * @return the list
+     * @throws DBException the db exception
+     */
+    List<? extends SQLScriptElement> translate(@NotNull SQLScriptElement element) throws DBException;
```
Please specify nullability using an appropriate annotation

oksana-cherniavskaia replied:
fixed

ShadelessFox commented on plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Sql dialect ddl extension.
+ */
+public interface SQLDialectDDLExtension {
```
Do we really need this extension interface?

I suggest adding these methods to the `SQLDialect` interface and removing this extension.

If you want to keep it, please make it extend `SQLDialect`.

serge-rider replied:
Yes, we do. DDL extension may add tens of new function. It doesn't make sense for most dialects.
Especially for NoSQL ones.

serge-rider commented on plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java:
```diff
@@ -42,10 +43,10 @@
 
     private static final String[][] TSQL_BEGIN_END_BLOCK = new String[][]{
-        {SQLConstants.BLOCK_BEGIN, SQLConstants.BLOCK_END}
+            {SQLConstants.BLOCK_BEGIN, SQLConstants.BLOCK_END}
```
Continuous indent = 4 spaces.

oksana-cherniavskaia replied:
fixed

serge-rider commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java:
```diff
@@ -0,0 +1,71 @@
+/*
```
Why do we need a new  class jsut for one function?
Let's move it to SQLTranslatorHandler class.

oksana-cherniavskaia replied:
SQLTranslatorHandler located in org.jkiss.dbeaver.ui.app.standalone.cli and seems to be not suitable place for our translation as we do not work with cli stuff.

oksana-cherniavskaia replied:
Moved to SQLQueryTranslator

Commit:
CB-2534 Support sql server and oracle as internal db: apply fixes after review.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java:
```diff
@@ -40,31 +41,35 @@
 import org.jkiss.utils.CommonUtils;
 
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
```
We do use `*` imports for 5 and more imports from the same package.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java:
```diff
@@ -190,7 +195,11 @@ public MultiValueInsertMode getDefaultMultiValueInsertMode() {
 
     @Override
-    public String getColumnTypeModifiers(@NotNull DBPDataSource dataSource, @NotNull DBSTypedObject column, @NotNull String typeName, @NotNull DBPDataKind dataKind) {
+    public String getColumnTypeModifiers(@NotNull DBPDataSource dataSource,
+                                         @NotNull DBSTypedObject column,
+                                         @NotNull String typeName,
+                                         @NotNull DBPDataKind dataKind
+    ) {
```
```suggestion
    public String getColumnTypeModifiers(
        @NotNull DBPDataSource dataSource,
        @NotNull DBSTypedObject column,
        @NotNull String typeName,
        @NotNull DBPDataKind dataKind
    ) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/parser/SQLScriptParser.java:
```diff
@@ -806,19 +806,19 @@ public static List<SQLScriptElement> parseScript(DBPDataSource dataSource, Strin
     }
 
-    public static SQLScriptElement parseQuery(SQLDialect dialect, DBPPreferenceStore preferenceStore, String sqlScriptContent, int cursorPosition) {
+    public static SQLScriptElement parseQuery(SQLDialect dialect, DBPPreferenceStore preferenceStore, String sqlScriptContent,
+                                              int cursorPosition) {
```
Please add nullability annotations

```suggestion
    public static SQLScriptElement parseQuery(
        SQLDialect dialect,
        DBPPreferenceStore preferenceStore,
        String sqlScriptContent,
        int cursorPosition
    ) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/parser/SQLScriptParser.java:
```diff
@@ -806,19 +806,19 @@ public static List<SQLScriptElement> parseScript(DBPDataSource dataSource, Strin
+    }
+
+    @NotNull
+    private static SQLParserContext prepareSqlParserContext(SQLDialect dialect, DBPPreferenceStore preferenceStore,
+                                                            String sqlScriptContent) {
```
Please add nullability annotations

```suggestion
    private static SQLParserContext prepareSqlParserContext(
        SQLDialect dialect,
        DBPPreferenceStore preferenceStore,
        String sqlScriptContent
    ) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:
```diff
@@ -56,5 +88,159 @@ public static String translateScript(
+     * @return the list
+     */
+    @NotNull
+    protected List<? extends SQLScriptElement> translateStatement(@NotNull SQLQuery query,
+                                                                  @NotNull Statement statement
```
```suggestion
    protected List<? extends SQLScriptElement> translateStatement(
        @NotNull SQLQuery query,
        @NotNull Statement statement
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Sql dialect ddl extension.
+ */
+public interface SQLDialectDDLExtension {
```
```suggestion
public interface SQLDialectDDLExtension extends SQLDialect {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return the auto increment keyword
+     */
+    String getAutoIncrementKeyword();
```
Please add nullability annotations

Commit:
CB-2534: Removes from SQLQueryTranslator wrong condition

Removes condition from SQLTranslator as supportsTableDropCascade from SQLDialect should not be taken into account on translating foreign keys

Commit:
CB-2534: SQLDialectDDLExtension methods renamed
